test/distributed/_composable/fsdp/test_fully_shard_init.py

```diff
@@ -554,6 +566,91 @@ def test_invalid_meta_device_init(self):
         with self.assertRaisesRegex(RuntimeError, error_regex):
             model(inp)
 
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_rank0_broadcast_meta_device_init(self):
+        model_args = ModelArgs(dropout_p=0.0)
+        # Assume we have a CPU full state dict on rank 0
+        if self.rank == 0:
+            torch.manual_seed(42)
+            ref_model = Transformer(model_args)
+            full_sd = ref_model.state_dict()
+            for param in full_sd.values():
+                self.assertEqual(param.device, torch.device("cpu"))
+
+        # Initialize the sharded model on meta device
+        fsdp_mesh = init_device_mesh("cuda", (self.world_size,))
+        with torch.device("meta"):
+            model = Transformer(model_args)
+        for module in model.modules():
+            if isinstance(module, TransformerBlock):
+                fully_shard(module, mesh=fsdp_mesh)
+        fully_shard(model, mesh=fsdp_mesh)
+        for param in model.parameters():
+            self.assertEqual(param.device, torch.device("meta"))
+
+        # Construct a sharded state dict from the rank 0 full state dict by
+        # broadcasting and sharding
+        meta_sharded_sd = model.state_dict()
+        sharded_sd = {}
+        if self.rank == 0:
+            self.assertEqual(len(meta_sharded_sd), len(full_sd))
+            self.assertEqual(list(meta_sharded_sd.keys()), list(full_sd.keys()))
+            for (param_name, full_param), sharded_meta_param in zip(
+                full_sd.items(), meta_sharded_sd.values()
+            ):
+                full_param = full_param.detach().cuda()
+                mesh = sharded_meta_param.device_mesh
+                dist.broadcast(full_param, src=0, group=mesh.get_group(0))
+                sharded_tensor = distribute_tensor(
+                    full_param, mesh, sharded_meta_param.placements
+                )
+                sharded_sd[param_name] = nn.Parameter(sharded_tensor)
```

Comment:
curious if buffers are sometimes included in state dict? for llama2, the buffer is RotaryPositionalEmbeddings.theta but it's not included in HF state dict